pkg/networkserver/downlink.go

```diff
@@ -696,6 +703,58 @@ func nonRetryableFixedPathGatewayError(err error) bool {
 	return errors.IsNotFound(err) || errors.IsDataLoss(err) || errors.IsFailedPrecondition(err)
 }
 
+type downlinkTarget interface {
+	Equal(downlinkTarget) bool
+	Schedule(context.Context, *ttnpb.DownlinkMessage, ...grpc.CallOption) (time.Duration, error)
+}
+
+type gatewayServerDownlinkTarget struct {
+	peer cluster.Peer
+}
+
+func (t *gatewayServerDownlinkTarget) Equal(target downlinkTarget) bool {
+	other, ok := target.(*gatewayServerDownlinkTarget)
+	if !ok {
+		return false
+	}
+	return other.peer == t.peer
+}
+
+func (t *gatewayServerDownlinkTarget) Schedule(ctx context.Context, msg *ttnpb.DownlinkMessage, callOpts ...grpc.CallOption) (time.Duration, error) {
+	conn, err := t.peer.Conn()
+	if err != nil {
+		return 0, err
+	}
+	res, err := ttnpb.NewNsGsClient(conn).ScheduleDownlink(ctx, msg, callOpts...)
+	if err != nil {
+		return 0, err
+	}
+	return res.Delay, nil
+}
+
+type packetBrokerDownlinkTarget struct {
+	peer cluster.Peer
+}
+
+func (t *packetBrokerDownlinkTarget) Equal(target downlinkTarget) bool {
+	_, ok := target.(*packetBrokerDownlinkTarget)
+	return ok
+}
+
+const packetBrokerScheduleDelay = 5 * time.Second
```

Comment:
shouldn't this be based of `infrastructureDelay`?

Reply:
Yeah good one. There would be a multiplier but makes sense to start with `infrastructureDelay`.